protocols/virtualdefund/virtualdefund_test.go

```diff
@@ -0,0 +1,200 @@
+package virtualdefund
+
+import (
+	"bytes"
+	"fmt"
+	"math/big"
+	"testing"
+
+	"github.com/statechannels/go-nitro/channel/state"
+	"github.com/statechannels/go-nitro/channel/state/outcome"
+	"github.com/statechannels/go-nitro/internal/testactors"
+	ta "github.com/statechannels/go-nitro/internal/testactors"
+	"github.com/statechannels/go-nitro/internal/testhelpers"
+	"github.com/statechannels/go-nitro/protocols"
+	"github.com/statechannels/go-nitro/types"
+)
+
+var alice = ta.Alice
+var bob = ta.Bob
+var irene = ta.Irene
+var allActors = []ta.Actor{alice, irene, bob}
+
+// makeOutcome creates an outcome allocating to alice and bob
+func makeOutcome(aliceAmount uint, bobAmount uint) outcome.SingleAssetExit {
+	return outcome.SingleAssetExit{
+		Allocations: outcome.Allocations{
+			outcome.Allocation{
+				Destination: alice.Destination(),
+				Amount:      big.NewInt(int64(aliceAmount)),
+			},
+			outcome.Allocation{
+				Destination: bob.Destination(),
+				Amount:      big.NewInt(int64(bobAmount)),
+			},
+		},
+	}
+}
+
+type testdata struct {
+	vFixed         state.FixedPart
+	vFinal         state.State
+	initialOutcome outcome.SingleAssetExit
+	finalOutcome   outcome.SingleAssetExit
+	paid           uint
+}
+
+// generateTestData generates some test data that can be used in a test
+func generateTestData() testdata {
+	vFixed := state.FixedPart{
+		ChainId:           big.NewInt(9001),
+		Participants:      []types.Address{alice.Address, irene.Address, bob.Address}, // A single hop virtual channel
+		ChannelNonce:      big.NewInt(0),
+		AppDefinition:     types.Address{},
+		ChallengeDuration: big.NewInt(45),
+	}
+
+	initialOutcome := makeOutcome(7, 2)
+	finalOutcome := makeOutcome(6, 3)
+	paid := uint(1)
+
+	vFinal := state.StateFromFixedAndVariablePart(vFixed, state.VariablePart{IsFinal: true, Outcome: outcome.Exit{finalOutcome}, TurnNum: FinalTurnNum})
+
+	return testdata{vFixed, vFinal, initialOutcome, finalOutcome, paid}
+}
+
+// signByOthers signs the state by every participant except my
+func signByOthers(my ta.Actor, signedState state.SignedState) state.SignedState {
+	if my.Role != 0 {
+		_ = signedState.Sign(&alice.PrivateKey)
+	}
+
+	if my.Role != 1 {
+		_ = signedState.Sign(&irene.PrivateKey)
+	}
+
+	if my.Role != 2 {
+		_ = signedState.Sign(&bob.PrivateKey)
+	}
+	return signedState
+}
+
+// assertStateSentToEveryone asserts that ses contains a message for every participant but from
+func assertStateSentToEveryone(t *testing.T, ses protocols.SideEffects, expected state.SignedState, from testactors.Actor) {
```

Comment:
I find this function with 5 levels of nesting hard to interpret. Any way to flatten out this function?

Reply:
I've flattened this a bit in https://github.com/statechannels/go-nitro/pull/522/commits/a079b5aad7e99c791a04c97c585359202f87af90 by breaking it into two functions.